File: app/src/main/java/org/odddev/fantlab/core/utils/BindingUtils.java
package org.odddev.fantlab.core.utils;

import android.databinding.BindingAdapter;
import android.databinding.BindingConversion;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.support.design.widget.TextInputLayout;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.DrawableRequestBuilder;
import com.bumptech.glide.Glide;

import org.odddev.fantlab.R;

/**
 * @author kenrube
 * @since 17.09.16
 */

public class BindingUtils {

    @BindingAdapter("onHomeOptionSelected")
    public static void bindOnClick(View view, final Runnable runnable) {
        view.setOnClickListener(v -> runnable.run());
    }

    @BindingAdapter("error")
    public static void bindError(TextInputLayout textInputLayout, final String error) {
        textInputLayout.setErrorEnabled(error != null);
        textInputLayout.setError(error);
    }

    @BindingAdapter({"srcUri", "isCircle", "placeholder"})
    public static void bindSrcUri(ImageView imageView, final String uri, final boolean isCircle,
                                  final Drawable placeholder) {
        DrawableRequestBuilder<String> builder = Glide
                .with(imageView.getContext())
                .load(uri)
                .placeholder(placeholder)
                .error(placeholder);
        if (isCircle) {
            builder.transform(new CircleTransform(imageView.getContext()));
        }
        builder.into(imageView);
    }

    @BindingAdapter("font")
    public static void bindFont(TextView textView, final String fontName) {
        String robotoMedium = textView.getContext().getString(R.string.roboto_medium);
        if (fontName.equals(robotoMedium) && Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            textView.setTypeface(Typeface.createFromAsset(
                    textView.getContext().getAssets(), "fonts/" + robotoMedium + ".ttf"));
        } else {
            textView.setTypeface(Typeface.create(fontName, Typeface.NORMAL));
        }
    }

    @BindingConversion
    public static int convertConditionToVisibility(final boolean condition) {
        return condition ? View.VISIBLE : View.GONE;
    }
}

File: app/src/main/kotlin/org/odddev/fantlab/core/utils/BindingUtils.kt
package org.odddev.fantlab.core.utils

import android.databinding.BindingAdapter
import android.databinding.BindingConversion
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.os.Build
import android.support.design.widget.TextInputLayout
import android.view.View
import android.widget.CompoundButton
import android.widget.ImageView
import android.widget.TextView

import com.bumptech.glide.Glide

import org.odddev.fantlab.R

object BindingUtils {

	@BindingAdapter("onClick")
	@JvmStatic
	fun bindOnClick(view: View, runnable: Runnable) {
		view.setOnClickListener { runnable.run() }
	}

	@BindingAdapter("onCheck")
	@JvmStatic
	fun bindOnCheck(button: CompoundButton, runnable: Runnable) {
		button.setOnCheckedChangeListener { _, checked -> if (checked) runnable.run() }
	}

	@BindingAdapter("error")
	@JvmStatic
	fun bindError(textInputLayout: TextInputLayout, error: String?) {
		textInputLayout.isErrorEnabled = error != null
		textInputLayout.error = error
	}

	@BindingAdapter("srcUri", "isCircle", "placeholder")
	@JvmStatic
	fun bindSrcUri(imageView: ImageView, uri: String, isCircle: Boolean,
				   placeholder: Drawable) {
		val builder = Glide
				.with(imageView.context)
				.load(uri)
				.placeholder(placeholder)
				.error(placeholder)
		if (isCircle) {
			builder.transform(CircleTransform(imageView.context))
		}
		builder.into(imageView)
	}

	@BindingAdapter("srcUri")
	@JvmStatic
	fun bindSrcUri(imageView: ImageView, uri: String) {
		Glide
				.with(imageView.context)
				.load(uri)
				.into(imageView)
	}

	@BindingAdapter("font")
	@JvmStatic
	fun bindFont(textView: TextView, fontName: String) {
		val robotoMedium = textView.context.getString(R.string.roboto_medium)
		if (fontName == robotoMedium && Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
			textView.typeface = Typeface.createFromAsset(
					textView.context.assets, "fonts/$robotoMedium.ttf")
		} else {
			textView.typeface = Typeface.create(fontName, Typeface.NORMAL)
		}
		// todo организовать кеш шрифтов
	}

	@BindingConversion
	@JvmStatic
	fun convertConditionToVisibility(condition: Boolean) = if (condition) View.VISIBLE else View.GONE
}
